src/naming/Naming_AST.ml

```diff
@@ -964,9 +973,20 @@ let resolve lang prog =
         else
           Common.save_excursion env.in_type true (fun () ->
               super#visit_type_ venv x)
-      (* TODO: we should intercept also V.kstmt and especially
-       * create new blocks for For, If with complex init_condition.
-       *)
+
+      (* TODO: support other types of statements that create block scopes. *)
+      method! visit_stmt venv x =
+        match x.s with
+        | If (tok, Cond e, s1, s2_opt) when has_block_scope lang ->
+            self#visit_tok venv tok;
+            self#visit_expr venv e;
+            with_new_block_scope env.names (fun () -> self#visit_stmt venv s1);
```

Comment:
Long time since I looked carefully into `Naming_AST`, but I'm assuming these new scopes will only affect `VarDecl`s and not `Assign`s.

Reply:
Correct. Double-checked and it only affects VarDef.